File: js/2493.js
const fs = require('fs');

const inputs = fs.readFileSync('example.txt').toString().trim().split('\n');

const N = +inputs[0];
const arr = inputs[1].split(' ').map(Number);

const stack = [];
const ans = [];

for (let i = 0; i < N; i++) {
  if (stack.length === 0) ans.push(0);
  else {
    if (arr[stack[stack.length - 1] - 1] < arr[i]) {
      while (stack.length && arr[stack[stack.length - 1] - 1] < arr[i]) {
        stack.pop();
      }
      ans.push(stack.length === 0 ? 0 : stack[stack.length - 1]);
    } else {
      ans.push(stack[stack.length - 1]);
    }
  }
  stack.push(i + 1);
}

console.log(ans.join(' '));

File: c++/2493.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> arr;
vector<int> stack;

int main() {
  freopen("example.txt", "r", stdin);

  int N;
  cin >> N;

  for (int i = 0; i < N; i++) {
    int a;
    cin >> a;
    arr.push_back(a);
  }

  for (int i = 0; i < N; i++) {
    if (stack.size() == 0) {
      cout << 0 << ' ';
    } else {
      if (arr[stack.back() - 1] < arr[i]) {
        while (stack.size() != 0 && arr[stack.back() - 1] < arr[i]) {
          stack.pop_back();
        }
        cout << (stack.size() == 0 ? 0 : stack.back()) << ' ';
      } else {
        cout << stack.back() << ' ';
      }
    }
    stack.push_back(i + 1);
  }

  return 0;
}